### PR TITLE
Contribution guidelines

### PR DESCRIPTION
**Pull request by timss**

### CODE REVIEW

**Comment by timss** [comment]
Somewhat relevant: #168 

And yeah, I already broke my own guidelines :-)

What do you think @gkatsev, following what I said in https://github.com/vimperator/vimperator-labs/issues/168#issuecomment-192901735 should we move stuff over from HACKING to CONTRIBUTING.md?

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+      frob();
+  else
+      unfrob();
+  ```
+
+- Prefer lambda-style functions where suitable:
+
+  ```javascript
+  // Good
+  list.filter(function (elem) elem.good != elem.BAD);
```
Comment (gkatsev) [inline CONTRIBUTING.md]: this is no longer valid. "Expression closures" are being removed plus non-standard.
`function (elem) returnValue` that is.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+  list.forEach(function (elem) { window.alert(elem); });
+
+  // Bad
+  list.filter(function (elem) { return elem.good != elem.BAD });
+  list.forEach(function (elem) window.alert(elem));
```
Comment (gkatsev) [inline CONTRIBUTING.md]: ditto here.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+- No braces for one-line conditional statements:
+
+  ```javascript
+  if (foo)
+      frob();
+  else
+      unfrob();
+  ```
+
+- Prefer lambda-style functions where suitable:
```
Comment (gkatsev) [inline CONTRIBUTING.md]: Actually, this whole section should probably be removed...

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+
+  ```javascript
+  // Good
+  function () {}
+
+  // Bad
+  function() {}
+  ```
+
+- Prefer the use of `let` over `var` i.e. only use `var` when required.
```
Comment (gkatsev) [inline CONTRIBUTING.md]: We now have `const` which I guess would be preferable.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+
+- Documentation comment blocks use `/** ... */` Wrap these lines at 80
+  characters.
+
+- Only wrap lines if it makes the code obviously clearer. Lines longer than 132
+  characters should probably be broken up rather than wrapped anyway.
+
+- Use UNIX new lines (`\n`), not windows (`\r\n`) or old Mac ones (`\r`).
+
+- Use `Iterator` or `Array#forEach` to iterate over arrays.
```
Comment (gkatsev) [inline CONTRIBUTING.md]: This section needs to be updated.
We should prefer array iterator functions like `forEach` or `map` but not use array comprehensions or non-standard iterators like `for each (let i in a)`. `for (let i of a)` could be use as that's a new way of doing it.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+
+In general: Just look at the existing source code!
+
+We try to target experienced JavaScript developers who do not necessarily need
+to have a good understanding of Vimperator's source code, nor necessarily
+understand in-depth concepts of other languages like Lisp or Python. Therefore,
+the coding style should feel natural to any JavaScript developer. Of course,
+this does not mean, you have to avoid all new JavaScript features like list
+comprehension or generators. Use them, when they make sense, but don't use them
+when the resulting code is hard to read.
```
Comment (gkatsev) [inline CONTRIBUTING.md]: Can we add a note below this paragraph (perhaps in bold or italics) to only use standards compliant javascript?

File: CONTRIBUTING.md (hunk 4 above)
Comment (timss) [inline CONTRIBUTING.md]: All right. What about [developer.mozilla.org - Features not compliant with the ES6 specification][1]?

If it doesn't matter, I was thinking maybe:

> - Adhere to modern JavaScript variable declarations, i.e. prefer `const` 
>   over `let`, and `let` over `var`.
> 
>   For more details, see:
> 
>   - https://hacks.mozilla.org/2015/07/es6-in-depth-let-and-const/
>   - https://developer.mozilla.org/en/New_in_JavaScript_1.7#Block_scope_with_let

[1]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/New_in_JavaScript/ECMAScript_6_support_in_Mozilla#Features_not_compliant_with_the_ES6_specification

File: CONTRIBUTING.md (hunk 3 above)
Comment (timss) [inline CONTRIBUTING.md]: Ok.

File: CONTRIBUTING.md (hunk 5 above)
Comment (timss) [inline CONTRIBUTING.md]: Sounds good, but could you possibly formulate an updated version? Not really my domain of expertise and I don't want to add any unnecessary ambiguity :\

File: CONTRIBUTING.md (hunk 6 above)
Comment (timss) [inline CONTRIBUTING.md]: :+1:

File: CONTRIBUTING.md (hunk 5 above)
Comment (gkatsev) [inline CONTRIBUTING.md]: Um... change it to

```
Prefer Array iterator functions `Array#forEach` `Array#map` over loops and array comprehensions
```

And remove lines 198 through 210. We can update it later.

File: CONTRIBUTING.md (hunk 4 above)
Comment (gkatsev) [inline CONTRIBUTING.md]: I guess it's ok to just leave it and we can get it updated later.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+
+- Only wrap lines if it makes the code obviously clearer. Lines longer than 132
+  characters should probably be broken up rather than wrapped anyway.
+
+- Use UNIX new lines (`\n`), not windows (`\r\n`) or old Mac ones (`\r`).
+
+- Prefer Array iterator functions `Array#forEach` and `Array#map` over loops
+  and array comprehensions.
+
+  The exceptions to this rule are for objects with `__iterator__` set, and for
```
Comment (gkatsev) [inline CONTRIBUTING.md]: oops, this line should also be removed.

Comment (timss) [inline CONTRIBUTING.md]: Ah shit, thought you intentionally left it behind. There we go.

**Pull request merged**